visvis2/_wrappers.py

```diff
@@ -0,0 +1,204 @@
+import ctypes
+
+import wgpu
+import numpy as np
+
+
+def array_from_shader_type(spirv_type):
+    """ Get a numpy array object from a SpirV type from python-shader.
+    """
+    return np.asarray(spirv_type())
+
+
+# todo: can this be generic enough, keeping the GPU bits out / optional?
+
+# todo: Support for updating unmapped data. Use something like updateRange to do subBufferUpdate
+
+
+class BaseBufferWrapper:
+    """ A base buffer wrapper that can be implemented for numpy, ctypes arrays,
+    or any other kind of array.
+    """
+
+    def __init__(self, data=None, nbytes=None, usage=None, mapped=False):
+        self._data = data
+        if nbytes is not None:
+            self._nbytes = int(nbytes)
+        elif data is not None:
+            self._nbytes = self._nbytes_from_data(data)
+        else:
+            raise ValueError("Buffer must be instantiated with either data or nbytes.")
+        if isinstance(usage, int):
+            self._usage = usage
+        elif isinstance(usage, str):
+            usages = usage.upper().replace(",", " ").replace("|", " ").split()
+            assert usages
+            self._usage = 0
+            for usage in usages:
+                self._usage |= getattr(wgpu.BufferUsage, usage)
+        else:
+            self._usage = wgpu.BufferUsage.STORAGE | wgpu.BufferUsage.VERTEX
+            # raise ValueError("BufferWrapper usage must be int or str.")
+
+        self._mapped = bool(mapped)
+        self._dirty = True
+        self._gpu_buffer = None  # Set by renderer
+
+    @property
+    def data(self):
+        """ The data that is a view on the data. Can be None if the
+        data only exists on the GPU.
+        Note that this array can be replaced, so get it via this property.
+        """
+        # todo: maybe this class should not store _data if the data is not mapped?
+        return self._data
+
+    @property
+    def nbytes(self):
+        """ Get the number of bytes in the buffer.
+        """
+        return self._nbytes
+
+    @property
+    def usage(self):
+        """ The buffer usage flags (as an int).
+        """
+        return self._usage
+
+    @property
+    def mapped(self):
+        """ Whether the data is mapped. Mapped data can be updated in-place
+        to change the data on the GPU.
+        """
+        return self._mapped
+
+    @property
+    def dirty(self):
+        """ Whether the buffer is dirty (needs to be processed by the renderer).
+        """
+        return self._dirty
+
+    @property
+    def strides(self):
+        """ Stride info (as a tuple).
+        """
+        return self._get_strides()
```

Comment:
I'm trying to see if this class can support interleaved buffers. Can it be done using strides and shape only? I think you would need more configuration options...

Reply:
> Does this class support interleaved buffers? 

Not really at the moment, but I think it can be made so.

The wgpu-renderer needs `array_stride` for vertex buffers (which is `BufferWrapper.strides[0]`.